include/nuttx/mm/circ_buf.h

```diff
@@ -0,0 +1,308 @@
+
+/* Note about locking: There is no locking required until only one reader
+ * and one writer is using the circular buffer.
+ * For multiple writer and one reader there is only a need to lock the
+ * writer. And vice versa for only one writer and multiple reader there is
+ * only a need to lock the reader.
+ */
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdbool.h>
+#include <sys/types.h>
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* This structure describes circular buffer */
+
+struct circ_buf_s
+{
+  FAR void *base;
+  size_t    size;
+  size_t    head;
+  size_t    tail;
+  bool      external;
+};
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
+#undef EXTERN
+#if defined(__cplusplus)
+#define EXTERN extern "C"
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+/****************************************************************************
+ * Name: circ_buf_init
+ *
+ * Description:
+ *   Initialize a circular buffer.
+ *
+ * Input Parameters:
+ *   circ  - Address of the circular buffer to be used.
+ *   base  - A pointer to circular buffer's internal buffer. It can provide
+ *           by called context because sometimes the creation of buffer
+ *           is speical or preallocated, eg: DMA buffer.
+ *           It also can alloc by this api if base is NULL.
+ *   bytes - The size of the internal buffer.
+ *
+ * Returned Value:
+ *   Zero on success; A negated errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+int circ_buf_init(FAR struct circ_buf_s *circ,
+                  FAR void *base, size_t bytes);
+
+/****************************************************************************
+ * Name: circ_buf_resize
+ *
+ * Description:
+ *   Resize a circular buffer (change buffer size).
+ *
+ * Input Parameters:
+ *   circ  - Address of the circular buffer to be used.
+ *   bytes - The size of the internal buffer.
+ *
+ * Returned Value:
+ *   Zero on success; A negated errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+int circ_buf_resize(FAR struct circ_buf_s *circ, size_t bytes);
+
+/****************************************************************************
+ * Name: circ_buf_uninit
+ *
+ * Description:
+ *   Free the circular buffer.
+ *
+ * Input Parameters:
+ *   circ  - Address of the circular buffer to be used.
+ ****************************************************************************/
+
+void circ_buf_uninit(FAR struct circ_buf_s *circ);
```

Comment:
Would `destroy` or `free`  be a better name and give the notion of de-allocation

Reply:
The init would follow. 
`create` - `destroy`
`allocate` - `free`

Reply:
> 
> 
> The init would follow.
> `create` - `destroy`
> `allocate` - `free`

Hello, Davids, I define struct circ_buf_s instance in other device driver and call circ_buf_init to initialize, it feels better to init on the semantic.

Reply:
> 
> 
> Looks good, just few minor comments everywhere. Mostly concerned about uneeded memory use.
> 
> I'm not a fan of the naming with an underscore and abbreviation (circ_buf), would probably be better if cbuffer or ringbuffer would be used. But this is just aesthetic comment.
 
I just refer to linux naming convention, it may be better.  https://github.com/torvalds/linux/blob/master/include/linux/circ_buf.h <smile>

Reply:
No, you should follow the NuttX naming conventions as specified in the Coding Standard:  https://cwiki.apache.org/confluence/display/NUTTX/Coding+Standard#funcname

Personal preference is irrelevant.

Reply:
But I think circ_buf_s already follow the guide. Anyway, what's name you prefer? circ_buf, ringbuffer or cbuffer. We can change to what you want.

Reply:
> 
> 
> But I think circ_buf_s already follow the guide. Anyway, what's name you prefer? circ_buf, ringbuffer or cbuffer. We can change to what you want.

There is a more complete description of naming conventions for OS internal functions in this wiki page:  https://cwiki.apache.org/confluence/display/NUTTX/Naming+of+OS+Internal+Functions

Reply:
@patacongo basically, circ_buf API follow the suggested style:
```
<subsystem>_<verb>_<object>[_<units>]
```
circ_buf is <sbusystem>, init... is <verb>, the difference is that API is simple enough to remove the object.

Reply:
I think what "breaks the rule" is the underscore. IMHO it would look better with any name which does not include an underscore (eg: circbuf, cbuffer, ringbuf, etc.). Anyway, I'll let you decide on it.

Reply:
okay. Done!